packages/protocol/contracts/signal/SignalService.sol

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: MIT
+//
+// ╭━━━━╮╱╱╭╮╱╱╱╱╱╭╮╱╱╱╱╱╭╮
+// ┃╭╮╭╮┃╱╱┃┃╱╱╱╱╱┃┃╱╱╱╱╱┃┃
+// ╰╯┃┃┣┻━┳┫┃╭┳━━╮┃┃╱╱╭━━┫╰━┳━━╮
+// ╱╱┃┃┃╭╮┣┫╰╯┫╭╮┃┃┃╱╭┫╭╮┃╭╮┃━━┫
+// ╱╱┃┃┃╭╮┃┃╭╮┫╰╯┃┃╰━╯┃╭╮┃╰╯┣━━┃
+// ╱╱╰╯╰╯╰┻┻╯╰┻━━╯╰━━━┻╯╰┻━━┻━━╯
+pragma solidity ^0.8.9;
+
+import "../common/EssentialContract.sol";
+import "../common/IHeaderSync.sol";
+import "../libs/LibBlockHeader.sol";
+import "../libs/LibTrieProof.sol";
+import "./ISignalService.sol";
+
+contract SignalService is ISignalService, EssentialContract {
+    using LibBlockHeader for BlockHeader;
+
+    struct SignalProof {
+        BlockHeader header;
+        bytes proof;
+    }
+
+    /// @dev Initializer to be called after being deployed behind a proxy.
+    function init(address _addressManager) external initializer {
+        EssentialContract._init(_addressManager);
+    }
+
+    function sendSignal(bytes32 signal) public returns (bytes32 storageSlot) {
+        require(signal != 0, "B:signal");
+
+        storageSlot = getSignalSlot(msg.sender, signal);
+        assembly {
+            sstore(storageSlot, 1)
+        }
+    }
+
+    function isSignalSent(
+        address app,
+        bytes32 signal
+    ) public view returns (bool) {
+        require(app != address(0), "B:app");
+        require(signal != 0, "B:signal");
+
+        bytes32 k = getSignalSlot(app, signal);
+        uint256 v;
+        assembly {
+            v := sload(k)
+        }
+        return v == 1;
+    }
+
+    function isSignalReceived(
+        uint256 srcChainId,
+        address app,
+        bytes32 signal,
+        bytes calldata proof
+    ) public view returns (bool) {
+        require(srcChainId != block.chainid, "B:srcChainId");
+        require(app != address(0), "B:app");
+        require(signal != 0, "B:signal");
+
+        SignalProof memory sp = abi.decode(proof, (SignalProof));
+        bytes32 syncedHeaderHash = IHeaderSync(resolve("taiko"))
+            .getSyncedHeader(sp.header.height);
+
+        if (
+            syncedHeaderHash == 0 ||
+            syncedHeaderHash != sp.header.hashBlockHeader()
+        ) {
+            return false;
+        }
+
+        return
+            LibTrieProof.verify({
+                stateRoot: sp.header.stateRoot,
+                addr: resolve(srcChainId, "signal_service"),
+                key: getSignalSlot(app, signal),
+                value: bytes32(uint256(1)),
+                mkproof: sp.proof
+            });
+    }
+
+    function getSignalSlot(
+        address app,
+        bytes32 signal
+    ) public pure returns (bytes32) {
```

Comment:
@davidtaikocha Do we want to give public access to getSignalSlot? In my understanding we can restrict to private/internal, not sure if dApps need it since they can just check for a signal using isSignalReceived() params

Reply:
But to get the actual Merkel proof, do they need to know the actual slot? @davidtaikocha

Reply:
we need to know the actual slot, lets keep this function public.